Stellar playground enhanced manual close

rokopt opened the pull request:
# Description

The `stellar-core` component of #2695 .

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests *(but new tests must be added before this PR comes out of draft)*
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md) **(N/A)**

jonjove commented on src/herder/HerderImpl.cpp:
```diff
@@ -1008,7 +1073,7 @@ HerderImpl::triggerNextLedger(uint32_t ledgerSeqToTrigger,
 
```
Are we using `slotIndex` here correctly in the case when we're closing with a distant `ledgerSeq`? (Wasn't allowed to comment on the line I wanted, but it's right above this)

rokopt replied:
Hmm, I'm not sure.  From my first reading I thought that when we're in `RUN_STANDALONE` mode, it sufficed for `slotIndex` to be increasing, but maybe it does have to match the new ledger sequence number.

I've changed it to be the latter because that now looks more plausible to me, but I'm still not sure what are all the ways that `slotIndex` is used.  This is probably a question for @MonsieurNicolas : if an enhanced manual ledger close increases the sequence number by more than `1`, should the `slotIndex` be one greater than the old sequence number, or equal to the new, even greater sequence number?  Or does it require subtler handling than either of those extremes?

rokopt replied:
I'm hoping that this has been addressed in the latest implementation, which reverts to using `triggerNextLedger()` only to increment the sequence number by `1` (it currently uses `advanceLedgerPointers()` to do the big jump).

jonjove commented on src/main/ApplicationImpl.cpp:
```diff
@@ -587,7 +587,7 @@ ApplicationImpl::manualClose(std::map<std::string, std::string> const& retMap,
         return;
     }
-    throw std::runtime_error("Enhanced manual close not implemented yet");
+    mHerder->triggerEnhancedManualCloseLedger(retMap, retStr);
```
This doesn't seem to do all the same things as the other `manualClose`. Ideally these should share the same code path (after all, if `retMap.empty()` then they should be _identical_). You can either remove one, or have one invoke the other. 

Furthermore, as with `Herder::triggerEnhancedManualCloseLedger` this interface is not appropriate here. Handling the key-value pairs should be happening higher up in the call-stack.

rokopt replied:
I've moved the key-value-pair handling to `CommandHandler.cpp`, and used the resulting restructuring to unify the `manualClose()` and `triggerNextLedger()` variants into single interfaces.

jonjove reviewed (changes requested):
Some general issues:

- Commit time-stamps are out of order (although GitHub seems to be showing them in the right order... did that behavior change?)
- fdb9fd6f doesn't seem to have anything to do with this PR, so it should be moved to a separate PR

Overall this seems like a good start. I had a couple of comments about how various interfaces were designed. One thing I was pretty unsure about was modifying `triggerNextLedger`, although I didn't see a better path. That function is already too long. Not sure what to do with that.

rokopt commented:
> * Commit time-stamps are out of order (although GitHub seems to be showing them in the right order... did that behavior change?)

Okay, I've rebased them into order (with `--ignore-date`).  I'm not sure whether GitHub changed anything.

jonjove commented:
If you liked the old commit order (which I was fine with), then you can also just rewrite the dates. I use a command like `git rebase --force-rebase --exec 'sleep 2' --exec 'git commit --amend --no-edit --date=now' <ref>` and I believe other people on our team have similar commands.

rokopt commented:
> * [fdb9fd6](https://github.com/stellar/stellar-core/commit/fdb9fd6ffbee628df48d7bcd9077bb5ef2682d0a) doesn't seem to have anything to do with this PR, so it should be moved to a separate PR

Okay, I thought I was going to depend on that, but I ended up not doing so.  I've rebased it away.  It already has PR #2626 .

rokopt commented on src/main/ApplicationUtils.cpp:
```diff
@@ -39,6 +39,16 @@ runWithConfig(Config cfg)
+        LOG(ERROR)
+            << "Starting stellar-core in ENHANCED_MANUAL_CLOSE mode requires "
+               "MANUAL_CLOSE to be set, NODE_IS_VALIDATOR to be set, and "
+               "RUN_STANDALONE to be unset";
```
Good catch.  Fixed.

rokopt commented:
> One thing I was pretty unsure about was modifying `triggerNextLedger`, although I didn't see a better path. That function is already too long. Not sure what to do with that.

Same here!

rokopt commented on src/main/ApplicationImpl.cpp:
```diff
@@ -578,6 +578,18 @@ ApplicationImpl::manualClose()
+ApplicationImpl::manualClose(std::map<std::string, std::string> const& retMap,
+                             std::string& retStr)
+{
+    if (!mConfig.ENHANCED_MANUAL_CLOSE)
```
I ended up doing it the way I did primarily because it means that if we ever decide to change the conditions under which `ENHANCED_MANUAL_CLOSE` is allowed to be set, we'll only have to change those checks, and not have to remember to change anything here.

Paranoia about such things isn't bad, though.  Still, if we _do_ end up changing the conditions, it could be to make them _more_ restrictive, in which case there could be the risk in forgetting to update them here that we later lead someone who notices the inconsistency to decide that the old, more lax requirements are the right ones.  Overall I think I favor leaving it and trusting ourselves to have done the validation correctly when we first set the flag.

jonjove replied:
I'm not sure you understood quite what I meant (not that I necessarily disagree with your decision, I just want to make sure we're speaking the same language). I'm talking about purely internal usage, like
```
Config cfg;
cfg.ENHANCED_MANUAL_CLOSE = true;
...
```
The checks that you added are only in `runWithConfig` which is only used via the command line.

rokopt replied:
Oh, I see.  Yes, I did fail to gather that.

I think I am inclined to let purely internal users (potentially) shoot themselves in the foot arbitrarily, but it's not a strong preference so I would be fine with going the other way if you'd rather.

rokopt commented:
> If you liked the old commit order (which I was fine with), then you can also just rewrite the dates. I use a command like `git rebase --force-rebase --exec 'sleep 2' --exec 'git commit --amend --no-edit --date=now' <ref>` and I believe other people on our team have similar commands.

I _think_ I preserved the commit order (I believe I'd have had at least one conflict if I hadn't), but all the commits ended up within the same second.  I do like how that command introduces a visible difference; I'll copy that one down.

MonsieurNicolas commented on src/herder/HerderImpl.cpp:
```diff
@@ -949,17 +1006,25 @@ HerderImpl::triggerNextLedger(uint32_t ledgerSeqToTrigger,
-        ctValidityOffset(nextCloseTime) == std::chrono::milliseconds::zero();
-
-    if (!isCtValid)
+    auto const enhancedManualClose = mApp.getConfig().ENHANCED_MANUAL_CLOSE;
```
I do not like this:
1. either we have full fidelity with `triggerNextLedger` (so no need to diverge from the existing code), or
2. we hook things up at the `ledgerClose` layer and aim to get partial fidelity
  * we're bypassing SCP but when there is a single value, this may not be that much of a big deal and the goal is to force the close time to take on a specific value anyways

"1" seems to be the best, but for that we need https://github.com/stellar/stellar-core/issues/1760
* any other way will require special handling (what you're doing here, and more) as the close time specified may not be compatible with the current date (when using non virtual time), and we need values to be valid during the SCP round (the part that you missed) - there might be even more special cases, and we don't want to go chase this stuff.
* a potential downside with this approach is that the "close time" specified may or may not be the actual time in the resulting ledger (as we'd be using all rules)

"2" will require extracting parts of `triggerNextLedger` so that we can share the code that calculates the candidate value, but instead of using SCP we would just invoke the close ledger code directly.

Both of those approaches require some special code that ensures that lcl is the "right one" when the ledger seqnum jumps ahead, otherwise SCP won't close that ledger and ledger close will crash (because it increments the ledger number by 1).

rokopt replied:
I thought that enforcing `RUN_STANDALONE` would allow us to set the sequence number forward arbitrarily since there would be no nodes proposing different ones.  I didn't realize that it was automatic that each close just incremented the number.

I _think_ we want advancing the sequence number to be one of the things that we're letting people test, right?  Are you saying we'll need to change the sequence number in the header in `triggerNextLedger()` if it jumps ahead because of a manual request, to satisfy SCP?

Are you also saying that we should force the use of virtual time when MANUAL_CLOSE is specified (issue #1760) as part of the work for this issue?

rokopt replied:
I'm now changing the sequence number in a separate modification before `triggerNextLedger()`, and I am indeed using `VIRTUAL_TIME` when `MANUAL_CLOSE` is specified, as suggested by issue #1760 .

MonsieurNicolas commented on src/main/ApplicationImpl.cpp:
```diff
@@ -399,20 +399,26 @@ ApplicationImpl::start()
         LOG(ERROR) << "Starting stellar-core with METADATA_OUTPUT_STREAM "
-                      "requires NODE_IS_VALIDATOR to be unset";
-        throw std::invalid_argument("NODE_IS_VALIDATOR is set");
+                      "requires NODE_IS_VALIDATOR to be unset or "
```
I'm not sure why we log before throwing, I'm pretty sure we log the exception anyways, so logging here seems redundant

rokopt replied:
Good question.  It was like that already; do you want me to change it?

rokopt replied:
Done.

MonsieurNicolas commented on src/main/ApplicationImpl.cpp:
```diff
@@ -399,20 +399,26 @@ ApplicationImpl::start()
     }
 
-    if (mConfig.METADATA_OUTPUT_STREAM != "" && mConfig.NODE_IS_VALIDATOR)
+    if (mConfig.METADATA_OUTPUT_STREAM != "" && mConfig.isNetworkedValidator())
```
it seems that this is the only place in the entire code base that we use `isNetworkedValidator`, I think that adding a variable with that name here (and using it for both conditions) avoids introducing a new concept to the code base

rokopt replied:
Alright, I've done that.

MonsieurNicolas commented on src/main/ApplicationUtils.cpp:
```diff
@@ -39,6 +39,15 @@ runWithConfig(Config cfg)
         }
     }
 
+    if (cfg.ENHANCED_MANUAL_CLOSE &&
```
see my other comment on the new flag: the right checks are already performed if we don't introduce this new config element, so we can remove this

rokopt replied:
I've removed `isStandaloneValidator()` as well.

rokopt replied:
And now I've removed `ENHANCED_MANUAL_CLOSE` altogether.

MonsieurNicolas commented on src/main/CommandHandler.cpp:
```diff
@@ -167,6 +167,15 @@ CommandHandler::fileNotFound(std::string const& params, std::string& retStr)
+
+    if (!retMap.empty())
+    {
+        mApp.manualClose(retMap, retStr);
```
we should really share the code with manual close (I think this will become evident if we use virtual time).
I would just expect the behavior to be additive: for example, if we need to "jump ahead" the ledger seq num, we would do that before performing the existing "manualClose" behavior, same thing with the close time.

rokopt replied:
If I understand what you're suggesting correctly, then this is done now.  `Application::manualClose()` is now a single interface with optional parameters for the sequence number and close time, and it's called in just one place from `CommandHandler::manualClose()`.

rokopt commented on src/herder/HerderImpl.cpp:
```diff
@@ -917,11 +918,67 @@ HerderImpl::setInSyncAndTriggerNextLedger()
+    {
+        setNumericParam(
+            ledgerSeq, "ledgerSeq",
+            [](std::string stringVal) { return stoul(stringVal); },
```
Also a good catch.  When I started writing that function, I meant to compensate for the lack of a `std::stou()`, but by the time I got there, I'd forgotten.  Fixed (I hope in a sane way).

MonsieurNicolas replied:
This code is redundant - if you do the parsing in the http handler, you can use things like `maybeParseParam` that performs the right checks

rokopt replied:
Thanks, `maybeParseParam()` is nice.  I'm using that now that I've moved the argument-parsing into `CommandHandler.cpp`.

rokopt commented on src/herder/HerderImpl.cpp:
```diff
@@ -917,11 +918,67 @@ HerderImpl::setInSyncAndTriggerNextLedger()
 
+void
+HerderImpl::triggerEnhancedManualCloseLedger(
+    std::map<std::string, std::string> const& retMap, std::string& retStr)
```
Makes sense.  I've moved the argument parsing into `CommandHandler.cpp`.

rokopt commented on docs/quick-reference.md:
```diff
@@ -78,6 +78,11 @@ The `manualclose` command allows to close a ledger, this is used when the instan
 configured with `MANUAL_CLOSE=true`.
 With this you can test various scenario where ledger boundaries could be important.
 
+Setting `ENHANCED_MANUAL_CLOSE=true` (along with `MANUAL_CLOSE=true`) further
```
Alright, I took it out.

I do still validate that we're in `RUN_STANDALONE` before allowing the specification of any parameters to the `manualclose` HTTP command.  Does that make sense?  I don't think we'd want one validator in a network to be able to try to push forward the close time or ledger sequence number arbitrarily, would we?

jonjove commented on src/main/ApplicationUtils.cpp:
```diff
@@ -40,7 +40,8 @@ runWithConfig(Config cfg)
 
     LOG(INFO) << "Starting stellar-core " << STELLAR_CORE_VERSION;
-    VirtualClock clock(VirtualClock::REAL_TIME);
+    VirtualClock clock(cfg.MANUAL_CLOSE ? VirtualClock::VIRTUAL_TIME
```
This doesn't seem to implement everything in #1760.

jonjove replied:
Ah I see, it's in the next commit. Probably should squash them.

rokopt replied:
Squashed.

jonjove commented on src/herder/HerderImpl.cpp:
```diff
@@ -949,8 +949,16 @@ HerderImpl::triggerNextLedger(optional<uint32_t> ledgerSeqToTrigger,
     // We pick as next close time the current time unless it's before the last
     // close time. We don't know how much time it will take to reach consensus
     // so this is the most appropriate value to use as closeTime.
+    auto const usingManualCloseWithoutExplicitCloseTime =
```
Nitpick: `bool` and `auto` are the same length, why not be explicit?

rokopt replied:
That sounds fine to me, though as it turns out, later changes have eliminated that code altogether. :)

jonjove commented on src/herder/HerderImpl.cpp:
```diff
@@ -949,8 +949,16 @@ HerderImpl::triggerNextLedger(optional<uint32_t> ledgerSeqToTrigger,
     // so this is the most appropriate value to use as closeTime.
+    auto const usingManualCloseWithoutExplicitCloseTime =
+        mApp.getConfig().MANUAL_CLOSE && !minCloseTime;
+    auto static const manualCloseDefaultPerLedgerTimeTIncrement =
```
I'm not really understanding this constant. The `closeTime` has units of seconds, so you can just add 5. You don't need to convert it to a `time_t` or a `time_point` (which is not a duration).

Nitpick: We don't generally use local `static` variables, and I don't see any serious need for it here. Also can we change the name? It's a bit of an eye-sore and doesn't help me understand what it means.

rokopt replied:
The code that implements the five-second default has been moved (to `ApplicationImpl::manualClose()`), and it no longer uses any conversions (the constant is just a `5`).

jonjove commented on src/herder/HerderImpl.cpp:
```diff
@@ -949,8 +949,16 @@ HerderImpl::triggerNextLedger(optional<uint32_t> ledgerSeqToTrigger,
+            std::chrono::system_clock::time_point(std::chrono::seconds{5}));
     uint64_t nextCloseTime =
-        VirtualClock::to_time_t(mApp.getClock().system_now());
+        usingManualCloseWithoutExplicitCloseTime
```
More generally, does this code even need to be in `triggerNextLedger` or could it be in `manualClose` (transmitted via the `minTime` parameter)?

rokopt replied:
It has, indeed, since been moved out of `triggerNextLedger()` and into `manualClose()`.

jonjove commented on src/herder/HerderImpl.cpp:
```diff
@@ -1033,7 +1034,16 @@ HerderImpl::triggerNextLedger(optional<uint32_t> ledgerSeqToTrigger,
+                "Allowing sequence number jump without MANUAL_CLOSE");
+        }
         slotIndex = *ledgerSeqToTrigger;
+        LedgerTxn updateLastSeqnoToPredecessor(mApp.getLedgerTxnRoot());
```
Nitpick: perhaps just `ltx` (that's the name we conventionally use everywhere) unless you have some compelling reason to have a long detailed name?

rokopt replied:
The code that did this has moved; it's called `ltx` there now.

jonjove reviewed (commented):
A few new comments, mostly cosmetic. Squashing the MANUAL_CLOSE commits into one would probably add clarity for reviewers.

MonsieurNicolas commented on src/herder/HerderImpl.cpp:
```diff
@@ -914,13 +918,15 @@ HerderImpl::setInSyncAndTriggerNextLedger()
 // uses the state in LedgerManager to derive a starting position
 void
-HerderImpl::triggerNextLedger(uint32_t ledgerSeqToTrigger,
+HerderImpl::triggerNextLedger(optional<uint32_t> ledgerSeqToTrigger,
```
It looks that from the conversation https://github.com/stellar/stellar-core/pull/2696#discussion_r483277469 you went with the "virtual time" option.

In this case, you don't actually need to (or want to) change `triggerNextLedger` that should just work "out of the box" as to provide full fidelity; in fact I don't even think that you need to make any modification to herder and just make sure that you setup the "initial state" properly:

* if a close time is specified by `manualclose`, you should just "jump time ahead" (see `setCurrentVirtualTime`)
* if a ledger seq num is specified (and it requires "jumping ahead" from a ledger point of view), this should be done outside of `triggerNextLedger` by directly adjusting lcl

rokopt replied:
Yes, you're right, once I went to `VIRTUAL_TIME`, the seeming need to modify `triggerNextLedger()` melted away.  :)

The new implementation is  in `ApplicationImpl::manualClose()`, and it does use `setCurrentVirtualTime()` and explicit modification of `loadHeader().current()`.

MonsieurNicolas commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +558,109 @@ ApplicationImpl::joinAllThreads()
+        }
+        auto nextLedgerSeqToTrigger = startLedgerSeq;
+        bool lastClose;
+        do
```
This loop seems overly complicated, why not just closing one ledger like before?

I think all you need to make this work is ensure that we meet the conditions so that "herder" can trigger the ledger properly (what I mentioned twice already in different comments).

Glancing at the code, I think that all you need is to call `restoreSCPState` with the right `index`/`StellarValue` when you modify lcl (NB: from `StellarValue`, we only use the close time).

rokopt replied:
>This loop seems overly complicated, why not just closing one ledger like before?

> I think all you need to make this work is ensure that we meet the conditions so that "herder" can trigger the ledger properly (what I mentioned twice already in different comments).

It was because I hadn't figured out how to meet the conditions to skip ledger numbers yet.  I wanted to see if I could get _something_ working first.

> Glancing at the code, I think that all you need is to call restoreSCPState with the right index/StellarValue when you modify lcl (NB: from StellarValue, we only use the close time).

Currently I'm taking Jon's suggestion of using (a slight variation on) `LedgerManager::advanceLedgerPointers()`, and thus far it appears to be working, but I can try the `restoreSCPState()` approach if that's better (maybe with just `advanceLedgerPointers()` there's something that's supposed to be happening that isn't?).

rokopt replied:
> Glancing at the code, I think that all you need is to call restoreSCPState with the right index/StellarValue when you modify lcl (NB: from StellarValue, we only use the close time).

Although some tests were working before using just Jon's `advanceLedgerPointers()` suggestion, there is at least this interface-violating way of calling `restoreSCPState()` as well, in case that is also necessary:  c23a2d05d370d2a67ff4e43cf3c33df9b9b6eb2c

For now, I've left that in a separate commit for independent review, since I'm not sure whether it's necessary or correct, and even if it's both, it should probably be exposed via some new Herder-interface abstract method, though I'm not sure with exactly what parameters or what to call it.

rokopt commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +558,109 @@ ApplicationImpl::joinAllThreads()
+                                       "sequence number {} (minimum {})"),
+                            targetLedgerSeq, minLedgerSeq));
+        }
+        if (!mHerder->isTransactionQueueEmpty() &&
```
> Why do we need this check? Because we already use the normal code when triggering ledgers we'll do the right thing with any potential transaction that gets invalidated by the new close time / ledger seq

When we manually close a ledger, there might be more transactions in the queue than would fit in a single ledger.  So to skip ledgers in that state, we have to close ledgers one at a time until the queue is empty, to get the right semantics regarding which transactions get into which ledgers (which may have further consequences).

So if we want to skip ahead, we have to choose where to do the one-by-one closing to empty the transaction queue first:  the client (usually Horizon), or the core.  The core could do it, or it could return an error and leave it to the client.  I think that it'd be basically the same code in either place, and I gather it's a generally-agreed principle (it would certainly make sense to me) that, all other things being equal, complexity is better outside `stellar-core` than inside it, so I went with leaving it to the client.

I could stick the loop in `stellar-core` itself if you prefer.

MonsieurNicolas replied:
not really: this is an arbitrary rule that should be removed.

In a real deployment of stellar-core, the decisions made wrt the transaction queue by the local node are pretty much independent of whatever happens during consensus. In fact this kind of thing happens pretty often as nodes on the network are not closing ledgers at exactly the same time but still accept/flood transactions.

Like you said, we should leave the decision on what to simulate to the caller, and this rule gets in the way.

rokopt replied:
> In a real deployment of stellar-core, the decisions made wrt the transaction queue by the local node are pretty much independent of whatever happens during consensus. In fact this kind of thing happens pretty often as nodes on the network are not closing ledgers at exactly the same time but still accept/flood transactions.

This is what I had failed to understand:  we might close a bunch of ledgers without taking any transactions from our queue, as it's consensus that dictates which transactions get in.  I had been stuck in "standalone mode" myself, thinking that the node would deterministically take some given set of transactions from its queue on each close.

I've removed `isTransactionQueueEmpty()`.

rokopt commented:
I think that I've answered the pending review comments as of now, so that this is ready for another round of reviews.

MonsieurNicolas commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +550,123 @@ ApplicationImpl::joinAllThreads()
+            ltx.commit();
+            auto const& header =
+                getLedgerManager().getLastClosedLedgerHeader().header;
+            dynamic_cast<HerderImpl&>(getHerder())
```
yes we need to do this.
I'd suggest adding a new method to Herder to expose this in a slightly more sane way and with a name that yells "unsafe".

Maybe something like `forceSyncedState()` (no parameters) that calls `restoreSCPState` with data gathered from lcl.

rokopt replied:
I've given it the descriptive-but-hostile-sounding name `forceSCPStateIntoSyncWithLastClosedLedger()`, though I'll shorten it to `forceSyncedState()` (or whatever) if you prefer.

MonsieurNicolas commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +550,123 @@ ApplicationImpl::joinAllThreads()
+                VirtualClock::time_point(std::chrono::seconds(nextCloseTime)));
+        }
+        mHerder->triggerNextLedger(targetLedgerSeq, true);
+        while (getClock().crank(true) != 0 &&
```
This doesn't look right:
we should not call `crank` from here as this code is already being called from the main `crank` and I don't think we make any guarantees that this function is re-entrant.

Because `triggerNextLedger` posts an event, things should "just work", did you have any problems without this code?

rokopt replied:
> did you have any problems without this code?

No, it _does_ just work.  This was me not adequately understanding how `VIRTUAL_TIME` worked.  I thought we'd have to manually `crank` to process the event(s) posted by `triggerNextLedger()` and the rest of the SCP processing resulting therefrom.  I've removed the `crank` loop.

rokopt commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +550,123 @@ ApplicationImpl::joinAllThreads()
+                                       .getLastClosedLedgerHeader()
+                                       .header.scpValue.closeTime;
+        auto const now = VirtualClock::to_time_t(getClock().system_now());
+        uint32_t const defaultManualCloseTimeInterval = 5;
```
Thanks, I didn't know we had that.  I'm using it now.

rokopt commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +550,123 @@ ApplicationImpl::joinAllThreads()
+        auto const now = VirtualClock::to_time_t(getClock().system_now());
+        uint32_t const defaultManualCloseTimeInterval = 5;
+        auto nextCloseTime = lastCloseTime + 1;
+        auto const defaultNextCloseTime =
```
Good point, done.

jonjove reviewed (commented):
Once `manualClose` has been refactored, I'll  want to review again to make sure you're covering all the right edge cases.

rokopt commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +549,109 @@ ApplicationImpl::joinAllThreads()
-        mHerder->triggerNextLedger(mLedgerManager->getLastClosedLedgerNum() + 1,
-                                   true);
-        return true;
+        auto const startLedgerSeq = getLedgerManager().getLastClosedLedgerNum();
```
Good idea; I've factored out three private utility methods from `ApplicationImpl::manualClose()`.

rokopt commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +549,109 @@ ApplicationImpl::joinAllThreads()
+        auto const now = VirtualClock::to_time_t(getClock().system_now());
+        uint32_t const defaultManualCloseTimeInterval =
+            getConfig().getExpectedLedgerCloseTime().count();
+        auto nextCloseTime = lastCloseTime + 1;
```
It turns out it's messier than either:  `HerderImpl::triggerNextLedger()` uses `VirtualClock::to_time_t()` on the `closeTime`, so we need to do the same to validate against the last closed header's `closeTime`, and possibly to produce a `closeTime` derived from that.  Then we convert it back to a `system_time_point` to call `setCurrentVirtualTime()`.  So a `closeTime` that overflows is one that isn't stable under all that converting.  That's nine-point-something billion.  I think that the best way to check for that is to do that stability test itself, so that's what the code does now.

rokopt commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +549,109 @@ ApplicationImpl::joinAllThreads()
+        }
+        if (targetLedgerSeq != minLedgerSeq)
+        {
+            LOG(INFO) << "closing ledger with sequence number "
```
Yeah, I don't like that one either.  I've changed it.

rokopt commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +549,109 @@ ApplicationImpl::joinAllThreads()
-                                   true);
-        return true;
+        auto const startLedgerSeq = getLedgerManager().getLastClosedLedgerNum();
+        if (startLedgerSeq == std::numeric_limits<uint32_t>::max())
```
Right, I noticed in testing that the `scphistory` database crashes on such values because it stores the ledger number as an `INT`.  I've updated the validation to reflect that.

rokopt commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +549,109 @@ ApplicationImpl::joinAllThreads()
+        uint32_t const defaultManualCloseTimeInterval =
+            getConfig().getExpectedLedgerCloseTime().count();
+        auto nextCloseTime = lastCloseTime + 1;
+        if (nextCloseTime < now)
```
Done.

rokopt commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +549,109 @@ ApplicationImpl::joinAllThreads()
+        }
+        else
+        {
+            auto const defaultNextCloseTime =
```
The code to check for the weird overflow condition is now unified between those two cases.

rokopt commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +549,109 @@ ApplicationImpl::joinAllThreads()
+        {
+            auto const defaultNextCloseTime =
+                lastCloseTime + defaultManualCloseTimeInterval;
+            if (nextCloseTime < defaultNextCloseTime)
```
Done.

rokopt commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +549,109 @@ ApplicationImpl::joinAllThreads()
+                targetLedgerSeq, newLedgerSeq));
+        }
+        return fmt::format(
+            FMT_STRING("Closed ledgers up to sequence number {}"),
```
True.  I've changed it.

rokopt commented on src/main/test/CommandHandlerTests.cpp:
```diff
@@ -137,3 +139,106 @@ TEST_CASE("transaction envelope bridge", "[commandhandler]")
     }
 }
+
+TEST_CASE("manualclose", "[commandhandler]")
```
I've added tests for those.  (I still have some more to write, such as to validate that the combinations of config parameters that I expect to be rejected are indeed rejected, but I think those tests should just be additions; the existing test code is reviewable now, as far as I know.  I have validated the config combinations manually with `curl`.)

rokopt commented:
I've dealt with code review comments again, and squashed commits.

I've also added a separate commit, d04230d1384901759d88b5878f0465334fe16dc9 , after discovering that when I would start `stellar-core` with `MANUAL_CLOSE` and `RUN_STANDALONE`, it would consume 100% of CPU time, and the virtual "now" would increase continually.  That turned out to be because the Herder's rebroadcast timer would continually fire and re-set itself when in virtual time with the clock being continually cranked.

jonjove commented on src/herder/HerderImpl.cpp:
```diff
@@ -318,6 +318,10 @@ HerderImpl::broadcast(SCPEnvelope const& e)
 void
 HerderImpl::startRebroadcastTimer()
 {
+    if (mApp.getConfig().MANUAL_CLOSE && mApp.getConfig().RUN_STANDALONE)
```
More generally, why do we do this at all if we are running with `RUN_STANDALONE`? There is no one to receive the broadcast.

rokopt replied:
I wondered if it might be okay to change that regardless of `MANUAL_CLOSE`.  I've done so now.

rokopt replied:
Actually, disabling it when only `RUN_STANDALONE` is present breaks, at least, the "3 nodes 2 running threshold 2"/"Over loopback" acceptance test.  I've pushed a change to revert to the original version which disables the timer only if both `MANUAL_CLOSE` and `RUN_STANDALONE` are configured.

jonjove replied:
@MonsieurNicolas can you open an issue for that with some useful context about why it is like this? It strongly defies my expectations that we might need to broadcast when running standalone.

MonsieurNicolas replied:
Yeah @jonjove I opened https://github.com/stellar/stellar-core/issues/2719

jonjove reviewed (commented):
Getting very close now.

rokopt commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +550,177 @@ ApplicationImpl::joinAllThreads()
+    optional<uint32_t> const& userSpecifiedSeqNum)
+{
+    auto const startLedgerSeq = getLedgerManager().getLastClosedLedgerNum();
+    if (startLedgerSeq == std::numeric_limits<uint32_t>::max())
```
> If lcl.ledgerSeq > INT32_MAX, then haven't we already had an issue with the scphistory table?

Yes, we would have.

> Why not just check [ ... ]

I've now done effectively that, though I've broken it up into separate tests so as to provide different log messages for each.

rokopt commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +550,177 @@ ApplicationImpl::joinAllThreads()
+}
+
+uint32_t
+ApplicationImpl::targetManualCloseLedgerSeqNum(
```
I've broken the code up into more "paragraphs".  Please let me know if it reads well to you now.

rokopt commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +550,177 @@ ApplicationImpl::joinAllThreads()
+    uint64_t const lastCloseTime = getLedgerManager()
+                                       .getLastClosedLedgerHeader()
+                                       .header.scpValue.closeTime;
+    if (lastCloseTime == std::numeric_limits<uint64_t>::max())
```
I'm now using `std::numeric_limits<std::time_t>::max()`.

And there's more:

- The overflow that I was actually running into was not of `time_t`, but of `system_time_point` in units of `std::chrono::seconds`.  I'm now checking against that as well as against `std::numeric_limits<std::time_t>::max()`.

- `HerderImpl::triggerNextLedger()` does validation that involves adding `Herder::MAX_TIME_SLIP_SECONDS` to the candidate close time.  So I have to subtract that from the minimum limit imposed by the `time_t` and `system_time_point` types to obtain the limit against which I need to validate.  I'm now doing that, and testing for it in the `"manualclose"` `TEST_CASE` in `CommandHandlerTests.cpp`.

MonsieurNicolas commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +550,224 @@ ApplicationImpl::joinAllThreads()
+    }
+    else
+    {
+        uint64_t const defaultManualCloseTimeInterval =
```
just realized that this is a breaking change:
* existing `manualclose` users are going to see a clock from the 1970s instead of a current clock, this may break clients that are setting time bounds in their transactions
* a non breaking change would be to set `defaultNextCloseTime` to the current *real* clock's value (so calling `std::chrono::steady_clock::now()` directly)

rokopt replied:
> calling `std::chrono::steady_clock::now()` directly

Why `std::chrono::steady_clock::now()` as opposed to `std::chrono::system_clock::now()`?  The latter is what `VirtualClock::system_now()` uses in `REAL_TIME`, and `VirtualClock::system_now()` is what `HerderImpl::triggerNextLedger()` uses.

rokopt replied:
Done in 8f353399bbdac17b5645942260f5809e4cb7031f, though as I mentioned, I used `std::chrono::system_clock::now()`, not `std::chrono::steady_clock::now()`.  Please let me know if I'm misunderstanding something there.

MonsieurNicolas replied:
This is fine yes, I was worried that the clock would go backwards, but we're guarding against that now, so people would just have to wait a bit if this happens

MonsieurNicolas commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +550,224 @@ ApplicationImpl::joinAllThreads()
+    TimePoint const lastCloseTime = getLedgerManager()
+                                        .getLastClosedLedgerHeader()
+                                        .header.scpValue.closeTime;
+    if (lastCloseTime >= maxCloseTime)
```
Why do we care to compare `lastCloseTime` to the max? We should just check that the new time is within bounds and avoid duplicate code?
(this is true for both checks)

rokopt replied:
The salient part of the `lastCloseTime` comparison is with `std::numeric_limits<TimePoint>::max()` -- maybe I should just test for equality with that.  If I don't, then the `lastCloseTime + 1` later might overflow, and that happens before we can check the new time.  With `secondsSinceEpoch`, that's making sure that `VirtualClock::to_time_t(systemNow)` won't overflow, which as Jon pointed out, could be undefined behavior.  So I might be able to change at least the first check to be against a different maximum (but I figured that as long as I was checking it, I may as well check it against the strictest thing that I could, as there would be no point continuing if it failed that check anyway).  Can I actually avoid either of these somehow without risking either of those overflows?

MonsieurNicolas replied:
That's because you wrote the code that way, if we use the current clock you won't do any arithmetic on time and can just enforce that `nextCloseTime > `lcl.ct`.

rokopt replied:
> That's because you wrote the code that way, if we use the current clock you won't do any arithmetic on time and can just enforce that nextCloseTime > lcl.ct`.

Don't we also have to enforce that a manually-specified close time is greater than the _virtual_ system now, regardless of whether we use the system clock as a default, because `HerderImpl::triggerNextLedger()` is going to use the virtual time?

rokopt replied:
I _think_ I've made the kind of rearrangement that you are requesting in 4932dab679e9f802c211ee35dd8d1c83655bb6e5.  Please let me know whether that is what you were looking for.

MonsieurNicolas replied:
yes things look good

MonsieurNicolas commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +550,224 @@ ApplicationImpl::joinAllThreads()
+
+    if (userSpecifiedCloseTime)
+    {
+        if (*userSpecifiedCloseTime > maxCloseTime)
```
I think we should just move those checks before calling `setCurrentVirtualTime` (and tweak the wording a bit as it's not necessarily provided by the user) - that way we perform the checks on the validity of `nextCloseTime` regardless of default behavior or not.

rokopt replied:
I think that has also happened as part of 4932dab679e9f802c211ee35dd8d1c83655bb6e5.

MonsieurNicolas reviewed (changes requested):
Looks pretty close to me as well. I have a few final suggestions to cleanup/simplify the code.

rokopt commented on src/main/Application.h:
```diff
@@ -5,6 +5,8 @@
 
 #include "main/Config.h"
+#include "util/optional.h"
+#include "xdr/Stellar-ledger-entries.h"
```
`TimePoint`.

rokopt commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +550,224 @@ ApplicationImpl::joinAllThreads()
+        {
+            if (manualCloseTime)
+            {
+                throw std::logic_error("Specifying manual close time should "
```
I don't think so -- if any options had been specified without `RUN_STANDALONE` being configured, then `CommandHandler::manualClose()` should have caught it:

```
    if (!retMap.empty() && !mApp.getConfig().RUN_STANDALONE)
    {
        throw std::invalid_argument(
            "The 'manualclose' command accepts parameters only if the "
            "configuration includes RUN_STANDALONE=true.");
    }
```

But `ApplicationImpl::manualClose()` is in a different class, so I figured that it should make it clear to any other potential users that they must also avoid providing parameters if `RUN_STANDALONE` isn't configured.

MonsieurNicolas replied:
Oh didn't see that... Mmmm I guess you can just use asserts then, those error messages make it look like they're for the end user, not a core developer and calling into "not implemented" behavior.

rokopt replied:
Okay, I've changed them to `releaseAssert()`s.

rokopt commented on src/main/test/CommandHandlerTests.cpp:
```diff
@@ -137,3 +144,366 @@ TEST_CASE("transaction envelope bridge", "[commandhandler]")
+
+TEST_CASE("manualclose", "[commandhandler]")
+{
+    auto testManualCloseConfig = [](auto configure, auto issue) {
```
You're welcome.  Have another.  ;)  (I added a mini-stress test.)

MonsieurNicolas replied:
let's not do stress testing as part of our unit tests:
* our test suite is already slow enough
* if we ever care about performance of this particular scenario, we'll tackle it then, when we actually understand the sequence of inputs that are causing performance issues

rokopt replied:
Alright, that one's gone.

rokopt commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +550,224 @@ ApplicationImpl::joinAllThreads()
+            VirtualClock::system_time_point::duration::max())
+            .count();
+    TimePoint const maxTimeT = std::numeric_limits<std::time_t>::max();
+    TimePoint const maxCloseTime =
```
> So instead, we should just pick a sensible constant easy to document for maxCloseTime like 1/1/2200 00:00:00 (the c++ spec says "at least 292 years" for second resolution, so we're well within that range relative to epoch).

Our code seems to fall short of the standard -- the value that produces overflow, `9'223'371'977`, is in the year 2262.  (But that does at least mean that the beginning of 2200 would work.  That's `7'258'118'400`, GMT.)

rokopt replied:
(I'll impose the beginning-of-2200 limit, and add a `static_assert` that that doesn't exceed the value that would produce overflow in `HerderImpl::triggerNextLedger()`.)

rokopt replied:
This is done in 6d1b919c14ad649d127e86040869b5deb0ae2706.

MonsieurNicolas replied:
no we're within standard: 292 years is for duration, and epoch starts in 1970, so if you try to add 1970+292 you end up at 2262.

rokopt replied:
Oh, right, I forgot that meant 292 years from _epoch_, not from _now_.

rokopt commented on src/main/ApplicationImpl.cpp:
```diff
@@ -549,27 +550,224 @@ ApplicationImpl::joinAllThreads()
+        }
+    }
+
+    auto const nextCloseSystemTime = VirtualClock::from_time_t(nextCloseTime);
```
I've used the hardcoded constant as mentioned above.  I made the `releaseAssert` change in 0545209546bf8826bf1a3ec51d9b25a1e9e795e9.

rokopt commented on src/main/ApplicationUtils.cpp:
```diff
@@ -37,10 +39,24 @@ runWithConfig(Config cfg)
+            if (cfg.AUTOMATIC_MAINTENANCE_COUNT != 0 ||
+                cfg.AUTOMATIC_MAINTENANCE_PERIOD.count() != 0)
+            {
+                LOG(ERROR) << "Using MANUAL_CLOSE and RUN_STANDALONE together "
```
Done in 819e6cea355a5ba73e71a179d5f25976962fd6b0.

rokopt commented:
I've made several more changes (without squashing commits this time) in response to Nicolas's latest comments.

I also noticed a bug myself, and fixed it in 1b4eafccfc65899f7914fe18d0fc426ea2aa0dd0.

I think it's ready for re-review.

jonjove commented on src/main/ApplicationImpl.cpp:
```diff
@@ -552,27 +553,224 @@ ApplicationImpl::joinAllThreads()
+    }
+
+    auto const nextCloseSystemTime = VirtualClock::from_time_t(nextCloseTime);
+    if (VirtualClock::to_time_t(nextCloseSystemTime) != nextCloseTime)
```
I know you have already removed these checks, but I want to emphasize that they don't make that much sense to begin with. As I was saying in my previous comment, `time_t` can be signed so you simply _cannot_ overflow it under any circumstances. You can't try to retroactively catch it, because overflowing it is already undefined behavior.

As an analogy, the following code is wrong
```
bool doesAdditionOverflow(int x, int y)
{
    int z = x + y;
    return (z - y) != x;
}
```
because the undefined behavior has already occurred by the time you maybe noticed (and I say maybe, because nothing is guaranteed!).

There is no action required on this, so feel free to resolve it once you have read it.

rokopt replied:
Makes sense, thanks.  Resolving.

jonjove reviewed (commented):
Just some nitpicks and no-action-required comments left from me.

rokopt commented on src/main/ApplicationImpl.cpp:
```diff
@@ -552,27 +553,224 @@ ApplicationImpl::joinAllThreads()
+    {
+        throw std::invalid_argument(fmt::format(
+            FMT_STRING("Manually closed ledger sequence number ({}) can not be "
+                       "increased without overflowing database entry"),
```
Fixed.

rokopt commented on src/main/ApplicationImpl.cpp:
```diff
@@ -552,27 +553,224 @@ ApplicationImpl::joinAllThreads()
+        {
+            throw std::invalid_argument(fmt::format(
+                FMT_STRING("Manual close ledger sequence number {} too large "
+                           "for database (maximum {})"),
```
Fixed.

rokopt commented:
> Referencing specific hashes is generally not the most useful approach on GitHub, because that hash is no longer in this PR (at least GitHub still has a reference to it now).

I thought that in the short term it might save a bit of time in review, but I can avoid it in future, of course.

MonsieurNicolas commented on src/main/test/CommandHandlerTests.cpp:
```diff
@@ -137,3 +145,447 @@ TEST_CASE("transaction envelope bridge", "[commandhandler]")
+            }
+        }
+
+        SECTION("Large transaction queues may be emptied over the course of "
```
this test is getting way too deep into testing "everything that core does", we already have tests for the transaction queue, and this is just duplicating that so there is no real added test coverage, on the other hand we're making our test suite slower.

rokopt replied:
Okay, I've removed it.

rokopt commented:
I've squashed commits after addressing Nicolas's latest review comments.

MonsieurNicolas commented:
r+ b91dd1b360073ee153669696d9c0214f117db2b3

MonsieurNicolas commented:
r+ 023ec9969997a4ebc31054015fd32e1c6d2fd86a
